### PR TITLE
Add update checker

### PR DESCRIPTION
**Pull request by JRoy**
Adds an update checker to Essentials that will check for the latest version on startup, on player join (permission is `essentials.updatecheck`), or manually with `/ess version`.

On startup, the update checker will fetch build information from a resource generated at compile time and determine if the build is a dev or stable build. If it's a stable build, the update checker will only check for a new release; while a dev build will check for any new commits.

There are 6 different types of messages the update checker will return;
* Identical: The current build is the latest stable release or latest dev build. This message is only shown in the `/ess version` command.
* Behind: If the current build is stable, it's an entire stable build behind, otherwise it's one or more dev builds behind.
* Diverged: The current build was made from a branch other than `2.x` and is also one or more dev builds behind the latest commit on `2.x`.
* Diverged Latest: The current build was made from a branch other than `2.x` but is based on the latest commit from `2.x`.
* Unknown: The current build either has invalid build information or was customly built. This message is show everywhere but on player join.
* Error: There was an error while fetching the latest version information.

Update checks can be disabled using the `update-check` option in `config.yml`.

### CODE REVIEW

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+public final class EssentialsUpdateChecker {
+    private final static String REPO = "EssentialsX/Essentials";
+    private final static String BRANCH = "2.x";
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java]:
```suggestion
    private static final String REPO = "EssentialsX/Essentials";
    private static final String BRANCH = "2.x";
```
wh-

**Comment by JRoy** [comment]
~~TODO @mdcfe: Custom chart on bstats for branch name~~

File: Essentials/src/main/java/com/earth2me/essentials/Essentials.java [hunk 2]
```diff
@@ -490,7 +542,7 @@ public void reload() {
     @Override
     public List<String> onTabComplete(final CommandSender sender, final Command command, final String commandLabel, final String[] args) {
         return onTabCompleteEssentials(sender, command, commandLabel, args, Essentials.class.getClassLoader(),
-            "com.earth2me.essentials.commands.Command", "essentials.", null);
+                "com.earth2me.essentials.commands.Command", "essentials.", null);
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/Essentials.java]:
```suggestion
            "com.earth2me.essentials.commands.Command", "essentials.", null);
```
Unnecessary spacing change

File: Essentials/src/main/java/com/earth2me/essentials/Essentials.java [hunk 3]
```diff
@@ -603,10 +655,10 @@ public boolean onCommandEssentials(final CommandSender cSender, final Command co
 
             if (bSenderBlock != null) {
                 if (getSettings().logCommandBlockCommands()) {
-                    Bukkit.getLogger().log(Level.INFO, "CommandBlock at {0},{1},{2} issued server command: /{3} {4}", new Object[] {bSenderBlock.getX(), bSenderBlock.getY(), bSenderBlock.getZ(), commandLabel, EssentialsCommand.getFinalArg(args, 0)});
+                    Bukkit.getLogger().log(Level.INFO, "CommandBlock at {0},{1},{2} issued server command: /{3} {4}", new Object[]{bSenderBlock.getX(), bSenderBlock.getY(), bSenderBlock.getZ(), commandLabel, EssentialsCommand.getFinalArg(args, 0)});
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/Essentials.java]:
```suggestion
                    Bukkit.getLogger().log(Level.INFO, "CommandBlock at {0},{1},{2} issued server command: /{3} {4}", new Object[] {bSenderBlock.getX(), bSenderBlock.getY(), bSenderBlock.getZ(), commandLabel, EssentialsCommand.getFinalArg(args, 0)});
```

File: Essentials/src/main/java/com/earth2me/essentials/Essentials.java [hunk 4]
```diff
@@ -603,10 +655,10 @@ public boolean onCommandEssentials(final CommandSender cSender, final Command co
...
                 }
             } else if (user == null) {
-                Bukkit.getLogger().log(Level.INFO, "{0} issued server command: /{1} {2}", new Object[] {cSender.getName(), commandLabel, EssentialsCommand.getFinalArg(args, 0)});
+                Bukkit.getLogger().log(Level.INFO, "{0} issued server command: /{1} {2}", new Object[]{cSender.getName(), commandLabel, EssentialsCommand.getFinalArg(args, 0)});
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/Essentials.java]:
```suggestion
                Bukkit.getLogger().log(Level.INFO, "{0} issued server command: /{1} {2}", new Object[] {cSender.getName(), commandLabel, EssentialsCommand.getFinalArg(args, 0)});
```

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java [hunk 5]
```diff
@@ -420,6 +420,53 @@ public void run() {
...
+                                case ERROR: {
+                                    user.sendMessage(tl("versionErrorPlayer"));
+                                    break;
+                                }
+                                default: {
+                                    break;
+                                }
+                            }
+                        });
+                    }
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java]: Maybe move this into a `showJoinUpdateCheck` method or something?

File: Essentials/src/main/java/com/earth2me/essentials/metrics/MetricsWrapper.java [hunk 6]
```diff
@@ -38,6 +39,7 @@ public MetricsWrapper(final Plugin plugin, final int pluginId, final boolean inc
         checkForcedMetrics();
         addPermsChart();
         addEconomyChart();
+        addCustomChart(new Metrics.SimplePie("releaseBranch", EssentialsUpdateChecker::getVersionBranch));
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/metrics/MetricsWrapper.java]: Preferably put this in its own method for consistency

File: build.gradle [hunk 7]
```diff
@@ -88,8 +89,10 @@ subprojects {
     // Version Injection
     processResources {
         inputs.property('fullVersion', fullVersion)
```
Comment (mdcfe) [inline build.gradle:97]:
```suggestion
        // Always process resources if version string or git branch changes
        inputs.property('fullVersion', fullVersion)
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 8]
```diff
@@ -388,10 +389,63 @@ private void runVersion(final Server server, final CommandSource sender, final S
...
+            } else {
+                final EssentialsUpdateChecker.UpdateToken releaseToken = EssentialsUpdateChecker.getReleaseToken().join();
+                switch (releaseToken.getBranchStatus()) {
+                    case IDENTICAL: {
+                        sender.sendMessage(tl("versionReleaseLatest"));
+                        break;
+                    }
+                    case BEHIND: {
+                        sender.sendMessage(tl("versionReleaseNew", EssentialsUpdateChecker.getLatestRelease()));
+                        //TODO download link? (https://github.com/EssentialsX/Website/issues/26)
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java]: Pending adding a release page to the website, see EssentialsX/Website#29

**Review by mdcfe — CHANGES_REQUESTED** [review]

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+        final List<String> versionStr = new BufferedReader(new InputStreamReader(Objects.requireNonNull(EssentialsUpdateChecker.class.getClassLoader().getResourceAsStream("release")), StandardCharsets.UTF_8)).lines().collect(Collectors.toList());
+        if (versionStr.size() == 2) {
+            if (versionStr.get(0).matches("\\d+\\.\\d+\\.\\d+-dev\\+\\d\\d-[0-9a-f]{7,40}")) {
+                identifier = versionStr.get(0).split("-")[2];
+                dev = true;
+            } else {
+                identifier = versionStr.get(0);
+            }
+            branch = versionStr.get(1);
+        }
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java]:
```suggestion
        final InputStream inputStream = EssentialsUpdateChecker.class.getClassLoader().getResourceAsStream("release");
        if (inputStream != null) {
            final List<String> versionStr = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8)).lines().collect(Collectors.toList());
            if (versionStr.size() == 2) {
                if (versionStr.get(0).matches("\\d+\\.\\d+\\.\\d+-dev\\+\\d\\d-[0-9a-f]{7,40}")) {
                    identifier = versionStr.get(0).split("-")[2];
                    dev = true;
                } else {
                    identifier = versionStr.get(0);
                }
                branch = versionStr.get(1);
            }
        }
```

Afaik `Objects.requireNonNull` in the static block here will cause the class to not initialize, so EssentialsX won't load if the release file is missing.

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+        versionIdentifier = identifier;
+        versionBranch = branch;
+        devBuild = dev;
+    }
+
+    private EssentialsUpdateChecker() {
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java]: Should have a config option to disable - as it stands, this PR makes external connections without letting the user control them.

Probably worth making the methods in this class non-static, passing in `ISettings` here, and only triggering update checks if enabled in config.

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static com.earth2me.essentials.I18n.tl;
+
+public final class EssentialsUpdateChecker {
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java]: Possibly move to `com.earth2me.essentials.updatecheck.UpdateChecker`? Base package is cluttered as it is.

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+                        pendingReleaseFuture.complete(cachedRelease = fetchDistance(latestRelease, getVersionIdentifier()));
+                    } catch (JsonSyntaxException | NumberFormatException e) {
+                        e.printStackTrace();
+                        pendingReleaseFuture.complete(new UpdateToken(BranchStatus.ERROR));
+                    }
+                } catch (IOException e) {
+                    e.printStackTrace();
+                    pendingReleaseFuture.complete(new UpdateToken(BranchStatus.ERROR));
+                }
+                pendingReleaseFuture = null;
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java]: Why set to null here?

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+                lastFetchTime = System.currentTimeMillis();
+            }).start();
+        }
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java]:
```suggestion
        return pendingReleaseFuture;
        }
```
Without this, we would just return the cached result. On first run this is null and causes NPEs:

error

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+                e.printStackTrace();
+                return new UpdateToken(BranchStatus.ERROR);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+            return new UpdateToken(BranchStatus.ERROR);
+        }
+    }
+
+    public static String[] getVersionMethods(final boolean sendLatestMessage, final boolean verboseErrors) {
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java]: ...methods? 
```suggestion
    public static String[] getVersionMessages(final boolean sendLatestMessage, final boolean verboseErrors) {
```

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+                    return new String[] {tl(verboseErrors ? "versionError" : "versionErrorPlayer", EssentialsUpdateChecker.getBuildInfo())};
+                }
+                default: {
+                    return new String[] {};
+                }
+            }
+        }
+    }
+
+    public static class UpdateToken {
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsUpdateChecker.java]:
```suggestion
    public static class Version {
```

**Review by mdcfe — CHANGES_REQUESTED** [review]

**Review by mdcfe — DISMISSED** [review]

**Review by mdcfe — APPROVED** [review]

**Pull request merged**